src/plugins/kibana_utils/public/state_containers/create_state_container.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Subject } from 'rxjs';
+import { RecursiveReadonly } from '@kbn/utility-types';
+import {
+  PureTransitionsToTransitions,
+  PureTransition,
+  ReduxLikeStateContainer,
+  PureSelectorsToSelectors,
+} from './types';
+
+const $$observable = (typeof Symbol === 'function' && (Symbol as any).observable) || '@@observable';
+
+const freeze: <T>(value: T) => RecursiveReadonly<T> =
+  process.env.NODE_ENV !== 'production'
+    ? <T>(value: T): RecursiveReadonly<T> => {
+        if (!value) return value as RecursiveReadonly<T>;
+        if (typeof value === 'object') return Object.freeze({ ...value }) as RecursiveReadonly<T>;
+        else return value as RecursiveReadonly<T>;
+      }
+    : <T>(value: T) => value as RecursiveReadonly<T>;
+
+export const createStateContainer = <
+  State,
+  PureTransitions extends object,
+  PureSelectors extends object = {}
+>(
+  defaultState: State,
+  pureTransitions: PureTransitions,
+  pureSelectors: PureSelectors = {} as PureSelectors
+): ReduxLikeStateContainer<State, PureTransitions, PureSelectors> => {
+  const state$ = new Subject<RecursiveReadonly<State>>();
+  const container: ReduxLikeStateContainer<State, PureTransitions, PureSelectors> = {
+    state: freeze(defaultState),
+    get: () => container.state,
+    getState: () => container.state,
+    set: (state: State) => {
+      state$.next((container.state = freeze(state)));
+    },
+    state$,
```

Comment:
```
state$: state$.asObservable()
```

Otherwise .next() will be exposed and no type checking in `*.js` files to stop from using it

Reply:
It will still be exposed as `.source.next()` to JS users.

Reply:
Indeed :D 
But still I think `.asObservable()` is more honest in this case and better conveys intentions

Reply:
Using `BehaviourSubject` like you suggested, it turns out `.asObservable()` is not needed in that case.